# Clustering the Neighbourhoods of London and Paris

## Introduction


A Tale of Two cities, a novel written by Charles Dickens was set in London and Paris which takes place during the French Revolution. These cities were both happening then and now. A lot has changed over the years and we now take a look at how the cities have grown.

London and Paris are quite the popular tourist and vacation destinations for people all around the world. They are diverse and multicultural and offer a wide variety of experiences that is widely sought after. We try to group the neighbourhoods of London and Paris respectively and draw insights to what they look like now.

## Business Problem 

The aim is to help tourists choose their destinations depending on the experiences that the neighbourhoods have to offer and what they would want to have. This also helps people make decisions if they are thinking about migrating to London or Paris or even if they want to relocate neighbourhoods within the city. Our findings will help stakeholders make informed decisions and address any concerns they have including the different kinds of cuisines, provision stores and what the city has to offer.

##  Data Description

We require geolocation data for both London and Paris. Postal codes in each city serve as a starting point. Using Postal codes we use can find out the neighbourhoods, boroughs, venues and their most popular venue categories.

### London

To derive our solution, We scrape our data from https://en.wikipedia.org/wiki/List_of_areas_of_London

This wikipedia page has information about all the neighbourhoods, we limit it London.

    1. borough : Name of Neighbourhood
    2. town : Name of borough
    3. post_code : Postal codes for London.
This wikipedia page lacks information about the geographical locations. To solve this problem we use ArcGIS API

### ArcGIS API

ArcGIS Online enables you to connect people, locations, and data using interactive maps. Work with smart, data-driven styles and intuitive analysis tools that deliver location intelligence. Share your insights with the world or specific groups.

More specifically, we use ArcGIS to get the geo locations of the neighbourhoods of London. The following columns are added to our initial dataset which prepares our data.

    1. latitude : Latitude for Neighbourhood
    2. longitude : Longitude for Neighbourhood

### Paris

To derive our solution, We leverage JSON data available at https://www.data.gouv.fr/fr/datasets/r/e88c6fda-1d09-42a0-a069-606d3259114e

The JSON file has data about all the neighbourhoods in France, we limit it to Paris.

     1. postal_code : Postal codes for France
     2. nom_comm : Name of Neighbourhoods in France
     3. nom_dept : Name of the boroughs, equivalent to towns in France
     4. geo_point_2d : Tuple containing the latitude and longitude of the Neighbourhoods.

## Foursquare API Data

We will need data about different venues in different neighbourhoods of that specific borough. In order to gain that information we will use "Foursquare" locational information. Foursquare is a location data provider with information about all manner of venues and events within an area of interest. Such information includes venue names, locations, menus and even photos. As such, the foursquare location platform will be used as the sole data source since all the stated required information can be obtained through the API.

After finding the list of neighbourhoods, we then connect to the Foursquare API to gather information about venues inside each and every neighbourhood. For each neighbourhood, we have chosen the radius to be 500 meters.

The data retrieved from Foursquare contained information of venues within a specified distance of the longitude and latitude of the postcodes. The information obtained per venue as follows:

    1. Neighbourhood : Name of the Neighbourhood
    2. Neighbourhood Latitude : Latitude of the Neighbourhood
    3. Neighbourhood Longitude : Longitude of the Neighbourhood
    4. Venue : Name of the Venue
    5. Venue Latitude : Latitude of Venue
    6. Venue Longitude : Longitude of Venue
    7. Venue Category : Category of Venue
Based on all the information collected for both London and Paris, we have sufficient data to build our model. We cluster the neighbourhoods together based on similar venue categories. We then present our observations and findings. Using this data, our stakeholders can take the necessary decision.

## Methodology

We will be creating our model with the help of Python so we start off by importing all the required packages.

In [1]:
import pandas as pd
import requests
import numpy as np
import matplotlib.cm as cm
import matplotlib.colors as colors
import folium

# import k-means for the clustering stage
from sklearn.cluster import KMeans

The approach taken here is to explore each of the cities individually, plot the map to show the neighbourhoods being considered and then build our model by clustering all of the similar neighbourhoods together and finally plot the new map with the clustered neighbourhoods. We draw insights and then compare and discuss our findings.

## Exploring London

### Neighbourhoods of London

We begin to start collecting and refining the data needed for the our business solution to work.

### Data Collection

To get the neighbourhoods in london, we start by scraping the list of areas of london wiki page.

In [2]:
url_london = "https://en.wikipedia.org/wiki/List_of_areas_of_London"
wiki_london_url = requests.get(url_london)
wiki_london_url

<Response [200]>

Response 200 means that we are able to make the connection

In [3]:
wiki_london_data = pd.read_html(wiki_london_url.text)
wiki_london_data

[                                                   0
 0  Map all coordinates in "Category:Areas of Lond...
 1                 Download coordinates as: KML · GPX,
             Location                     London borough       Post town  \
 0         Abbey Wood              Bexley, Greenwich [7]          LONDON   
 1              Acton  Ealing, Hammersmith and Fulham[8]          LONDON   
 2          Addington                         Croydon[8]         CROYDON   
 3         Addiscombe                         Croydon[8]         CROYDON   
 4        Albany Park                             Bexley  BEXLEY, SIDCUP   
 ..               ...                                ...             ...   
 526         Woolwich                          Greenwich          LONDON   
 527   Worcester Park       Sutton, Kingston upon Thames  WORCESTER PARK   
 528  Wormwood Scrubs             Hammersmith and Fulham          LONDON   
 529          Yeading                         Hillingdon           HAYES   
 

In [4]:

wiki_london_data = wiki_london_data[1]
wiki_london_data

Location                     London borough       Post town  \
0         Abbey Wood              Bexley, Greenwich [7]          LONDON   
1              Acton  Ealing, Hammersmith and Fulham[8]          LONDON   
2          Addington                         Croydon[8]         CROYDON   
3         Addiscombe                         Croydon[8]         CROYDON   
4        Albany Park                             Bexley  BEXLEY, SIDCUP   
..               ...                                ...             ...   
526         Woolwich                          Greenwich          LONDON   
527   Worcester Park       Sutton, Kingston upon Thames  WORCESTER PARK   
528  Wormwood Scrubs             Hammersmith and Fulham          LONDON   
529          Yeading                         Hillingdon           HAYES   
530         Yiewsley                         Hillingdon    WEST DRAYTON   

    Postcode district Dial code OS grid ref  
0                 SE2       020    TQ465785  
1              W3, W4       020    TQ205805  
2                 CR0       020    TQ375645  
3                 CR0       020    TQ345665  
4           DA5, DA14       020    TQ478728  
..                ...       ...         ...  
526              SE18       020    TQ435795  
527               KT4       020    TQ225655  
528               W12       020    TQ225815  
529               UB4       020    TQ115825  
530               UB7       020    TQ063804  

[531 rows x 6 columns]

### Data Preprocessing

We remove the spaces in the column titles and then we add _ between words.

In [5]:

wiki_london_data.rename(columns=lambda x: x.strip().replace(" ", "_"), inplace=True)
wiki_london_data

Location                     London borough       Post_town  \
0         Abbey Wood              Bexley, Greenwich [7]          LONDON   
1              Acton  Ealing, Hammersmith and Fulham[8]          LONDON   
2          Addington                         Croydon[8]         CROYDON   
3         Addiscombe                         Croydon[8]         CROYDON   
4        Albany Park                             Bexley  BEXLEY, SIDCUP   
..               ...                                ...             ...   
526         Woolwich                          Greenwich          LONDON   
527   Worcester Park       Sutton, Kingston upon Thames  WORCESTER PARK   
528  Wormwood Scrubs             Hammersmith and Fulham          LONDON   
529          Yeading                         Hillingdon           HAYES   
530         Yiewsley                         Hillingdon    WEST DRAYTON   

    Postcode district Dial code OS_grid_ref  
0                 SE2       020    TQ465785  
1              W3, W4       020    TQ205805  
2                 CR0       020    TQ375645  
3                 CR0       020    TQ345665  
4           DA5, DA14       020    TQ478728  
..                ...       ...         ...  
526              SE18       020    TQ435795  
527               KT4       020    TQ225655  
528               W12       020    TQ225815  
529               UB4       020    TQ115825  
530               UB7       020    TQ063804  

[531 rows x 6 columns]

We see that few columns have no '_' between the words despite applying our function meaning that there are special characters

### Feature Selection

We need only the boroughs, Postal codes, Post town for further steps. We can drop the locations, dial codes and OS grid.

In [6]:
df1 = wiki_london_data.drop( [ wiki_london_data.columns[0], wiki_london_data.columns[4], wiki_london_data.columns[5] ], axis=1)

In [7]:
df1.head()

London borough       Post_town Postcode district
0              Bexley, Greenwich [7]          LONDON               SE2
1  Ealing, Hammersmith and Fulham[8]          LONDON            W3, W4
2                         Croydon[8]         CROYDON               CR0
3                         Croydon[8]         CROYDON               CR0
4                             Bexley  BEXLEY, SIDCUP         DA5, DA14

let's rename the Postcode district column and the london borough to something simpler.

In [8]:
df1.columns = ['borough','town','post_code']
df1

borough            town  post_code
0                Bexley, Greenwich [7]          LONDON        SE2
1    Ealing, Hammersmith and Fulham[8]          LONDON     W3, W4
2                           Croydon[8]         CROYDON        CR0
3                           Croydon[8]         CROYDON        CR0
4                               Bexley  BEXLEY, SIDCUP  DA5, DA14
..                                 ...             ...        ...
526                          Greenwich          LONDON       SE18
527       Sutton, Kingston upon Thames  WORCESTER PARK        KT4
528             Hammersmith and Fulham          LONDON        W12
529                         Hillingdon           HAYES        UB4
530                         Hillingdon    WEST DRAYTON        UB7

[531 rows x 3 columns]

Let's remove the Square brackets [ ] and numbers from the borough column

In [9]:
df1['borough'] = df1['borough'].map(lambda x: x.rstrip(']').rstrip('0123456789').rstrip('['))
df1

borough            town  post_code
0                Bexley, Greenwich           LONDON        SE2
1    Ealing, Hammersmith and Fulham          LONDON     W3, W4
2                           Croydon         CROYDON        CR0
3                           Croydon         CROYDON        CR0
4                            Bexley  BEXLEY, SIDCUP  DA5, DA14
..                              ...             ...        ...
526                       Greenwich          LONDON       SE18
527    Sutton, Kingston upon Thames  WORCESTER PARK        KT4
528          Hammersmith and Fulham          LONDON        W12
529                      Hillingdon           HAYES        UB4
530                      Hillingdon    WEST DRAYTON        UB7

[531 rows x 3 columns]

Take the dimension of the dataframe

In [10]:
df1.shape

(531, 3)

We currently have 531 records and 3 columns of our data. It's time to perform Feature Engineering

### Feature Engineering

We can only focusing on the neighbourhoods of London, so performing the changes

In [11]:
df1 = df1[df1['town'].str.contains('LONDON')]
df1

borough                    town post_code
0                Bexley, Greenwich                   LONDON       SE2
1    Ealing, Hammersmith and Fulham                  LONDON    W3, W4
6                              City                  LONDON       EC3
7                       Westminster                  LONDON       WC2
9                           Bromley                  LONDON      SE20
..                              ...                     ...       ...
521                       Redbridge                  LONDON  IG8, E18
522       Redbridge, Waltham Forest  LONDON, WOODFORD GREEN       IG8
525                          Barnet                  LONDON       N12
526                       Greenwich                  LONDON      SE18
528          Hammersmith and Fulham                  LONDON       W12

[308 rows x 3 columns]

In [12]:
df1.shape

(308, 3)

In [13]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 308 entries, 0 to 528
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   borough    308 non-null    object
 1   town       308 non-null    object
 2   post_code  308 non-null    object
dtypes: object(3)
memory usage: 9.6+ KB


## Geolocations of the London Neighbourhoods

### ArcGis API


We need to get the geographical co-ordinates for the neighbourhoods to plot out map. We will use the arcgis package to do so.

Arcgis doesn't have a limitation on the number of API calls made so it fits our use case perfectly.

In [14]:
import arcgis

In [15]:
from arcgis.geocoding import geocode
from arcgis.gis import GIS
gis = GIS()


Defining London arcgis geocode function to return latitude and longitude

In [16]:
def get_x_y_uk(address1):
   lat_coords = 0
   lng_coords = 0
   g = geocode(address='{}, London, England, GBR'.format(address1))[0]
   lng_coords = g['location']['x']
   lat_coords = g['location']['y']
   return str(lat_coords) +","+ str(lng_coords)

In [17]:
c = get_x_y_uk('SE2')
c

'51.492450000000076,0.12127000000003818'

Looks good, We Copy over the postal codes of london to pass it into the geolocator function that we just defined above

In [18]:
geo_coordinates_uk = df1['post_code']    
geo_coordinates_uk

0           SE2
1        W3, W4
6           EC3
7           WC2
9          SE20
         ...   
521    IG8, E18
522         IG8
525         N12
526        SE18
528         W12
Name: post_code, Length: 308, dtype: object

Passing postal codes of london to get the geographical co-ordinates

In [19]:
coordinates_latlng_uk = geo_coordinates_uk.apply(lambda x: get_x_y_uk(x))
coordinates_latlng_uk

0       51.492450000000076,0.12127000000003818
1        51.51324000000005,-0.2674599999999714
6       51.51200000000006,-0.08057999999994081
7       51.51651000000004,-0.11967999999995982
9       51.41009000000008,-0.05682999999993399
                        ...                   
521    51.589770000000044,0.030520000000024083
522      51.50642000000005,-0.1272099999999341
525     51.615920000000074,-0.1767399999999384
526      51.48207000000008,0.07143000000002075
528      51.50645000000003,-0.2369099999999662
Name: post_code, Length: 308, dtype: object

#### Latitude

Extracting the latitude from our previously collected coordinates

In [20]:
lat_uk = coordinates_latlng_uk.apply(lambda x: x.split(',')[0])
lat_uk

0      51.492450000000076
1       51.51324000000005
6       51.51200000000006
7       51.51651000000004
9       51.41009000000008
              ...        
521    51.589770000000044
522     51.50642000000005
525    51.615920000000074
526     51.48207000000008
528     51.50645000000003
Name: post_code, Length: 308, dtype: object

#### Longitude
Extracting the Longitude from our previously collected coordinates

In [21]:
lng_uk = coordinates_latlng_uk.apply(lambda x: x.split(',')[1])
lng_uk

0       0.12127000000003818
1       -0.2674599999999714
6      -0.08057999999994081
7      -0.11967999999995982
9      -0.05682999999993399
               ...         
521    0.030520000000024083
522     -0.1272099999999341
525     -0.1767399999999384
526     0.07143000000002075
528     -0.2369099999999662
Name: post_code, Length: 308, dtype: object

We now have the geographical co-ordinates of the London Neighbourhoods.

We proceed with Merging our source data with the geographical co-ordinates to make our dataset ready for the next stage

In [22]:
london_merged = pd.concat([df1,lat_uk.astype(float), lng_uk.astype(float)], axis=1)
london_merged.columns= ['borough','town','post_code','latitude','longitude']
london_merged

borough                    town post_code  \
0                Bexley, Greenwich                   LONDON       SE2   
1    Ealing, Hammersmith and Fulham                  LONDON    W3, W4   
6                              City                  LONDON       EC3   
7                       Westminster                  LONDON       WC2   
9                           Bromley                  LONDON      SE20   
..                              ...                     ...       ...   
521                       Redbridge                  LONDON  IG8, E18   
522       Redbridge, Waltham Forest  LONDON, WOODFORD GREEN       IG8   
525                          Barnet                  LONDON       N12   
526                       Greenwich                  LONDON      SE18   
528          Hammersmith and Fulham                  LONDON       W12   

     latitude  longitude  
0    51.49245    0.12127  
1    51.51324   -0.26746  
6    51.51200   -0.08058  
7    51.51651   -0.11968  
9    51.41009   -0.05683  
..        ...        ...  
521  51.58977    0.03052  
522  51.50642   -0.12721  
525  51.61592   -0.17674  
526  51.48207    0.07143  
528  51.50645   -0.23691  

[308 rows x 5 columns]

In [23]:
london_merged.dtypes

borough       object
town          object
post_code     object
latitude     float64
longitude    float64
dtype: object

#### Co-ordinates for London
Getting the geocode for London to help visualize it on the map

In [24]:
london = geocode(address='London, England, GBR')[0]
london_lng_coords = london['location']['x']
london_lat_coords = london['location']['y']
london_lng_coords

-0.1272099999999341

In [25]:
london_lat_coords

51.50642000000005

### Visualize the Map of London
To help visualize the Map of London and the neighbourhoods in London, we make use of the folium package.

In [26]:
# Creating the map of London
map_London = folium.Map(location=[london_lat_coords, london_lng_coords], zoom_start=12)
map_London

# adding markers to map
for latitude, longitude, borough, town in zip(london_merged['latitude'], london_merged['longitude'], london_merged['borough'], london_merged['town']):
    label = '{}, {}'.format(town, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_London)  
    
map_London

### Venues in London
To proceed with the next part, we need to define Foursquare API credentials.

Using Foursquare API, we are able to get the venue and venue categories around each neighbourhood in London.

In [27]:
CLIENT_ID = '3R1OMXTA3F5Z21QRLX3AA3AATWYAWCMZSZ1K00V35HC15ML2' 
CLIENT_SECRET = 'HJFI2UIUHXABQZSQU2NDUQ1NSDNJQ0IJEXDO0PTQTSVSSNQU'
VERSION = '20180605' # Foursquare API version

Defining a function to get the neraby venues in the neighbourhood. This will help us get venue categories which is important for our analysis.

In [28]:
LIMIT=100

def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius,
            LIMIT
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

Getting the venues in London

In [29]:
venues_in_London = getNearbyVenues(london_merged['borough'], london_merged['latitude'], london_merged['longitude'])

Bexley, Greenwich 
Ealing, Hammersmith and Fulham
City
Westminster
Bromley
Islington
Islington
Barnet
Enfield
Wandsworth
Southwark
City
Richmond upon Thames
Barnet
Islington
Wandsworth
Westminster
Bromley
Newham
Ealing
Westminster
Lewisham
Camden
Southwark
Tower Hamlets
Bexley
City
Lewisham
Greenwich
Tower Hamlets
Camden
Haringey
Tower Hamlets
Haringey
Barnet
Brent
Lambeth
Lewisham
Tower Hamlets
Kensington and Chelsea, Hammersmith and Fulham
Brent
Barnet
Barnet
Southwark
Tower Hamlets
Camden
Tower Hamlets
Waltham Forest
Newham
Islington
Richmond upon Thames
Lewisham
Camden
Westminster
Greenwich
Kensington and Chelsea
Barnet
Westminster
Lewisham
Waltham Forest
Hounslow, Ealing, Hammersmith and Fulham
Brent
Barnet
Lambeth, Wandsworth
Islington
Barnet
Merton
Barnet
Westminster
Barnet, Brent, Camden
Lewisham
Bexley
Haringey
Bromley
Tower Hamlets
Newham
Hackney
Islington
Southwark
Lewisham
Brent
Southwark
Ealing
Kensington and Chelsea
Wandsworth
Southwark
Barnet
Newham
Richmond upon Thames


Sampling our data

In [30]:
venues_in_London.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0  Bexley, Greenwich                 51.49245                  0.12127   
1  Bexley, Greenwich                 51.49245                  0.12127   
2  Bexley, Greenwich                 51.49245                  0.12127   
3  Bexley, Greenwich                 51.49245                  0.12127   
4  Bexley, Greenwich                 51.49245                  0.12127   

                              Venue     Venue Category  
0                      Lesnes Abbey      Historic Site  
1                       Sainsbury's        Supermarket  
2                              Lidl        Supermarket  
3  Abbey Wood Railway Station (ABW)      Train Station  
4                        Costcutter  Convenience Store

In [31]:
venues_in_London.shape

(10365, 5)

Wow, we have scraped together 10567 records for venues. This will definitely make the clustering interesting.

### Grouping by Venue Categories
We need to now see how many Venue Categories are there for further processing

In [32]:
venues_in_London.groupby('Venue Category').max()

Neighbourhood  Neighbourhood Latitude  \
Venue Category                                                        
Accessories Store               Westminster                51.51656   
Adult Boutique                    Islington                51.52969   
African Restaurant              Westminster                51.52587   
American Restaurant          Waltham Forest                51.61780   
Antique Shop                    Westminster                51.55506   
...                                     ...                     ...   
Wings Joint          Hammersmith and Fulham                51.54187   
Women's Store                   Westminster                51.55457   
Xinjiang Restaurant               Southwark                51.47480   
Yoga Studio                     Westminster                51.55457   
Zoo Exhibit                          Camden                51.53354   

                     Neighbourhood Longitude                        Venue  
Venue Category                                                             
Accessories Store                   -0.11968           James Smith & Sons  
Adult Boutique                      -0.08697  Sh! Women's Erotic Emporium  
African Restaurant                  -0.08808           Red Sea Restaurant  
American Restaurant                  0.02795                  Spielburger  
Antique Shop                        -0.11968     The London Silver Vaults  
...                                      ...                          ...  
Wings Joint                         -0.19795                     Wingmans  
Women's Store                       -0.11478           Vivien of Holloway  
Xinjiang Restaurant                 -0.09313                    Silk Road  
Yoga Studio                         -0.03558                    yogahaven  
Zoo Exhibit                         -0.14606                Penguin Beach  

[299 rows x 4 columns]

We can see 295 records, just goes to show how diverse and interesting the place is.

### One Hot Encoding
We need to Encode our venue categories to get a better result for our clustering

In [33]:
London_venue_cat = pd.get_dummies(venues_in_London[['Venue Category']], prefix="", prefix_sep="")
London_venue_cat

Accessories Store  Adult Boutique  African Restaurant  \
0                      0               0                   0   
1                      0               0                   0   
2                      0               0                   0   
3                      0               0                   0   
4                      0               0                   0   
...                  ...             ...                 ...   
10360                  0               0                   0   
10361                  0               0                   0   
10362                  0               0                   0   
10363                  0               0                   0   
10364                  0               0                   0   

       American Restaurant  Antique Shop  Arcade  Arepa Restaurant  \
0                        0             0       0                 0   
1                        0             0       0                 0   
2                        0             0       0                 0   
3                        0             0       0                 0   
4                        0             0       0                 0   
...                    ...           ...     ...               ...   
10360                    0             0       0                 0   
10361                    0             0       0                 0   
10362                    0             0       0                 0   
10363                    0             0       0                 0   
10364                    0             0       0                 0   

       Argentinian Restaurant  Art Gallery  Art Museum  ...  \
0                           0            0           0  ...   
1                           0            0           0  ...   
2                           0            0           0  ...   
3                           0            0           0  ...   
4                           0            0           0  ...   
...                       ...          ...         ...  ...   
10360                       0            0           0  ...   
10361                       0            0           0  ...   
10362                       0            0           0  ...   
10363                       0            0           0  ...   
10364                       0            0           0  ...   

       Vietnamese Restaurant  Warehouse Store  Whisky Bar  Wine Bar  \
0                          0                0           0         0   
1                          0                0           0         0   
2                          0                0           0         0   
3                          0                0           0         0   
4                          0                0           0         0   
...                      ...              ...         ...       ...   
10360                      0                0           0         0   
10361                      0                0           0         0   
10362                      0                0           0         0   
10363                      0                0           0         0   
10364                      0                0           0         0   

       Wine Shop  Wings Joint  Women's Store  Xinjiang Restaurant  \
0              0            0              0                    0   
1              0            0              0                    0   
2              0            0              0                    0   
3              0            0              0                    0   
4              0            0              0                    0   
...          ...          ...            ...                  ...   
10360          0            0              0                    0   
10361          0            0              0                    0   
10362          0            0              0                    0   
10363          0            0              0                    0   
10364          0            

Adding Neighbourhood into the mix.

In [34]:
London_venue_cat['Neighbourhood'] = venues_in_London['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [London_venue_cat.columns[-1]] + list(London_venue_cat.columns[:-1])
London_venue_cat = London_venue_cat[fixed_columns]

London_venue_cat.head()

Neighbourhood  Accessories Store  Adult Boutique  African Restaurant  \
0  Bexley, Greenwich                   0               0                   0   
1  Bexley, Greenwich                   0               0                   0   
2  Bexley, Greenwich                   0               0                   0   
3  Bexley, Greenwich                   0               0                   0   
4  Bexley, Greenwich                   0               0                   0   

   American Restaurant  Antique Shop  Arcade  Arepa Restaurant  \
0                    0             0       0                 0   
1                    0             0       0                 0   
2                    0             0       0                 0   
3                    0             0       0                 0   
4                    0             0       0                 0   

   Argentinian Restaurant  Art Gallery  ...  Vietnamese Restaurant  \
0                       0            0  ...                      0   
1                       0            0  ...                      0   
2                       0            0  ...                      0   
3                       0            0  ...                      0   
4                       0            0  ...                      0   

   Warehouse Store  Whisky Bar  Wine Bar  Wine Shop  Wings Joint  \
0                0           0         0          0            0   
1                0           0         0          0            0   
2                0           0         0          0            0   
3                0           0         0          0            0   
4                0           0         0          0            0   

   Women's Store  Xinjiang Restaurant  Yoga Studio  Zoo Exhibit  
0              0                    0            0            0  
1              0                    0            0            0  
2              0                    0            0            0  
3              0                    0            0            0  
4              0                    0            0            0  

[5 rows x 300 columns]

### Venue categories mean value
We will group the Neighbourhoods and calculate the mean venue categories value in each Neighbourhood

In [35]:
London_grouped = London_venue_cat.groupby('Neighbourhood').mean().reset_index()
London_grouped.head()

Neighbourhood  Accessories Store  Adult Boutique  \
0                 Barnet                0.0             0.0   
1  Barnet, Brent, Camden                0.0             0.0   
2                 Bexley                0.0             0.0   
3      Bexley, Greenwich                0.0             0.0   
4     Bexley, Greenwich                 0.0             0.0   

   African Restaurant  American Restaurant  Antique Shop  Arcade  \
0                 0.0             0.001805           0.0     0.0   
1                 0.0             0.000000           0.0     0.0   
2                 0.0             0.000000           0.0     0.0   
3                 0.0             0.000000           0.0     0.0   
4                 0.0             0.000000           0.0     0.0   

   Arepa Restaurant  Argentinian Restaurant  Art Gallery  ...  \
0               0.0                 0.00722          0.0  ...   
1               0.0                 0.00000          0.0  ...   
2               0.0                 0.00000          0.0  ...   
3               0.0                 0.00000          0.0  ...   
4               0.0                 0.00000          0.0  ...   

   Vietnamese Restaurant  Warehouse Store  Whisky Bar  Wine Bar  Wine Shop  \
0               0.001805              0.0         0.0       0.0        0.0   
1               0.000000              0.0         0.0       0.0        0.0   
2               0.000000              0.0         0.0       0.0        0.0   
3               0.000000              0.0         0.0       0.0        0.0   
4               0.000000              0.0         0.0       0.0        0.0   

   Wings Joint  Women's Store  Xinjiang Restaurant  Yoga Studio  Zoo Exhibit  
0          0.0            0.0                  0.0          0.0          0.0  
1          0.0            0.0                  0.0          0.0          0.0  
2          0.0            0.0                  0.0          0.0          0.0  
3          0.0            0.0                  0.0          0.0          0.0  
4          0.0            0.0                  0.0          0.0          0.0  

[5 rows x 300 columns]

Let's make a function to get the top most common venue categories

In [36]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

There are way too many venue categories, we can take the top 10 to cluster the neighbourhoods.

Creating a function to label the columns of the venue correctly

In [37]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

### Top venue categories
Getting the top venue categories in London

In [38]:
# create a new dataframe for London
neighborhoods_venues_sorted_london = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted_london['Neighbourhood'] = London_grouped['Neighbourhood']

for ind in np.arange(London_grouped.shape[0]):
    neighborhoods_venues_sorted_london.iloc[ind, 1:] = return_most_common_venues(London_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted_london.head()

Neighbourhood 1st Most Common Venue 2nd Most Common Venue  \
0                 Barnet           Coffee Shop                  Café   
1  Barnet, Brent, Camden                Bakery  Gym / Fitness Center   
2                 Bexley           Supermarket         Historic Site   
3      Bexley, Greenwich              Bus Stop           Sports Club   
4     Bexley, Greenwich            Supermarket     Convenience Store   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0         Grocery Store              Bus Stop                   Pub   
1     Convenience Store        Clothing Store        Hardware Store   
2              Platform     Convenience Store         Train Station   
3     Convenience Store           Golf Course         Historic Site   
4         Train Station              Platform         Historic Site   

        6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0          Italian Restaurant              Pharmacy           Supermarket   
1                 Supermarket           Zoo Exhibit   Filipino Restaurant   
2                 Golf Course              Bus Stop                  Park   
3  Construction & Landscaping                  Park               Daycare   
4                 Film Studio           Event Space               Exhibit   

        9th Most Common Venue 10th Most Common Venue  
0          Turkish Restaurant       Sushi Restaurant  
1                     Exhibit     Falafel Restaurant  
2  Construction & Landscaping   Fast Food Restaurant  
3           Food & Drink Shop            Flower Shop  
4          Falafel Restaurant         Farmers Market

## Model Building
### K Means
Let's cluster the city of london to roughly 5 to make it easier to analyze.

We use the K Means clustering technique to do so.

In [39]:
# set number of clusters
k_num_clusters = 5

London_grouped_clustering = London_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans_london = KMeans(n_clusters=k_num_clusters, random_state=0).fit(London_grouped_clustering)
kmeans_london

KMeans(algorithm='auto', copy_x=True, init='k-means++', max_iter=300,
       n_clusters=5, n_init=10, n_jobs=None, precompute_distances='auto',
       random_state=0, tol=0.0001, verbose=0)

### Labelling Clustered Data

In [40]:
kmeans_london.labels_

array([0, 0, 2, 4, 2, 3, 3, 3, 0, 3, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0, 0,
       0, 0, 0, 1, 0, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0])

In [41]:
neighborhoods_venues_sorted_london.insert(0, 'Cluster Labels', kmeans_london.labels_ +1)

Join London_merged with our neighbourhood venues sorted to add latitude & longitude for each of the neighborhood to prepare it for plotting

In [42]:
london_data = london_merged

london_data = london_data.join(neighborhoods_venues_sorted_london.set_index('Neighbourhood'), on='borough')

london_data.head()

borough    town post_code  latitude  longitude  \
0              Bexley, Greenwich   LONDON       SE2  51.49245    0.12127   
1  Ealing, Hammersmith and Fulham  LONDON    W3, W4  51.51324   -0.26746   
6                            City  LONDON       EC3  51.51200   -0.08058   
7                     Westminster  LONDON       WC2  51.51651   -0.11968   
9                         Bromley  LONDON      SE20  51.41009   -0.05683   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0               3           Supermarket     Convenience Store   
1               4         Grocery Store     Indian Restaurant   
6               1                 Hotel    Italian Restaurant   
7               1                 Hotel           Coffee Shop   
9               4           Supermarket  Fast Food Restaurant   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0         Train Station              Platform         Historic Site   
1         Train Station                  Park        Breakfast Spot   
6           Coffee Shop  Gym / Fitness Center                   Pub   
7        Sandwich Place                  Café                   Pub   
9                 Hotel         Grocery Store     Convenience Store   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0           Film Studio           Event Space               Exhibit   
1            Food Court     Food & Drink Shop           Flower Shop   
6              Wine Bar            Restaurant          Cocktail Bar   
7    Italian Restaurant             Hotel Bar               Theater   
9                  Park           Golf Course        Sandwich Place   

  9th Most Common Venue 10th Most Common Venue  
0    Falafel Restaurant         Farmers Market  
1           Flea Market            Fish Market  
6        Sandwich Place            Salad Place  
7            Restaurant       Sushi Restaurant  
9              Pharmacy   Gym / Fitness Center

In [43]:
london_data_nonan = london_data.dropna(subset=['Cluster Labels'])

### Visualizing the clustered neighbourhood
Let's plot the clusters

In [44]:
map_clusters_london = folium.Map(location=[london_lat_coords, london_lng_coords], zoom_start=12)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(london_data_nonan['latitude'], london_data_nonan['longitude'], london_data_nonan['borough'], london_data_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters_london)
        
map_clusters_london

### Examining our Clusters
Cluster 1

In [45]:
london_data_nonan.loc[london_data_nonan['Cluster Labels'] == 1, london_data_nonan.columns[[1] + list(range(5, london_data_nonan.shape[1]))]]

town  Cluster Labels 1st Most Common Venue  \
6                    LONDON               1                 Hotel   
7                    LONDON               1                 Hotel   
10                   LONDON               1           Coffee Shop   
12                   LONDON               1           Coffee Shop   
14           BARNET, LONDON               1           Coffee Shop   
..                      ...             ...                   ...   
521                  LONDON               1         Grocery Store   
522  LONDON, WOODFORD GREEN               1                  Café   
525                  LONDON               1           Coffee Shop   
526                  LONDON               1                   Pub   
528                  LONDON               1                   Pub   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
6      Italian Restaurant           Coffee Shop  Gym / Fitness Center   
7             Coffee Shop        Sandwich Place                  Café   
10                    Pub            Food Truck                  Café   
12                    Pub            Food Truck                  Café   
14                   Café         Grocery Store              Bus Stop   
..                    ...                   ...                   ...   
521                   Pub                  Café           Pizza Place   
522                 Hotel               Theater                   Pub   
525                  Café         Grocery Store              Bus Stop   
526         Grocery Store              Bus Stop           Coffee Shop   
528           Coffee Shop         Grocery Store                  Café   

     5th Most Common Venue       6th Most Common Venue 7th Most Common Venue  \
6                      Pub                    Wine Bar            Restaurant   
7                      Pub          Italian Restaurant             Hotel Bar   
10   Vietnamese Restaurant                        Park    Italian Restaurant   
12   Vietnamese Restaurant                        Park    Italian Restaurant   
14                     Pub          Italian Restaurant              Pharmacy   
..                     ...                         ...                   ...   
521                 Bakery                         Bar         Metro Station   
522                 Garden                       Plaza            Restaurant   
525                    Pub          Italian Restaurant              Pharmacy   
526      Indian Restaurant  Construction & Landscaping    Turkish Restaurant   
528              Gastropub                       Hotel                  Park   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
6            Cocktail Bar        Sandwich Place            Salad Place  
7                 Theater            Restaurant       Sushi Restaurant  
10   Gym / Fitness Center          Cocktail Bar                  Hotel  
12   Gym / Fitness Center          Cocktail Bar                  Hotel  
14            Supermarket    Turkish Restaurant       Sushi Restaurant  
..                    ...                   ...                    ...  
521    Seafood Restaurant            Restaurant            Coffee Shop  
522                Bakery     Outdoor Sculpture               Pharmacy  
525           Supermarket    Turkish Restaurant       Sushi Restaurant  
526           Golf Course     Fish & Chips Shop                   Park  
528       Thai Restaurant           Pizza Place     Italian Restaurant  

[279 rows x 12 columns]

Cluster 2

In [46]:
london_data_nonan.loc[london_data_nonan['Cluster Labels'] == 2, london_data_nonan.columns[[1] + list(range(5, london_data_nonan.shape[1]))]]

town  Cluster Labels 1st Most Common Venue  \
377  HARROW, STANMOREEDGWARE, LONDON               2                Bakery   

    2nd Most Common Venue    3rd Most Common Venue 4th Most Common Venue  \
377                   Gym  Health & Beauty Service  Ethiopian Restaurant   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
377           Event Space               Exhibit    Falafel Restaurant   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
377        Farmers Market  Fast Food Restaurant    Filipino Restaurant

Cluster 3

In [47]:
london_data_nonan.loc[london_data_nonan['Cluster Labels'] == 3, london_data_nonan.columns[[1] + list(range(5, london_data_nonan.shape[1]))]]

town  Cluster Labels 1st Most Common Venue  \
0                 LONDON               3           Supermarket   
45   BEXLEYHEATH, LONDON               3           Supermarket   
124               LONDON               3           Supermarket   
291       LONDON, SIDCUP               3           Supermarket   
505               LONDON               3           Supermarket   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0       Convenience Store         Train Station              Platform   
45          Historic Site              Platform     Convenience Store   
124         Historic Site              Platform     Convenience Store   
291         Historic Site              Platform     Convenience Store   
505         Historic Site              Platform     Convenience Store   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0           Historic Site           Film Studio           Event Space   
45          Train Station           Golf Course              Bus Stop   
124         Train Station           Golf Course              Bus Stop   
291         Train Station           Golf Course              Bus Stop   
505         Train Station           Golf Course              Bus Stop   

    8th Most Common Venue       9th Most Common Venue 10th Most Common Venue  
0                 Exhibit          Falafel Restaurant         Farmers Market  
45                   Park  Construction & Landscaping   Fast Food Restaurant  
124                  Park  Construction & Landscaping   Fast Food Restaurant  
291                  Park  Construction & Landscaping   Fast Food Restaurant  
505                  Park  Construction & Landscaping   Fast Food Restaurant

Cluster 4

In [48]:
london_data_nonan.loc[london_data_nonan['Cluster Labels'] == 4, london_data_nonan.columns[[1] + list(range(5, london_data_nonan.shape[1]))]]

town  Cluster Labels 1st Most Common Venue  \
1               LONDON               4         Grocery Store   
9               LONDON               4           Supermarket   
29   BECKENHAM, LONDON               4           Supermarket   
61              LONDON               4         Grocery Store   
69              LONDON               4         Grocery Store   
100             LONDON               4         Grocery Store   
127             LONDON               4           Supermarket   
137             LONDON               4         Grocery Store   
217             LONDON               4         Grocery Store   
260             LONDON               4         Grocery Store   
267             LONDON               4     Indian Restaurant   
269             LONDON               4         Grocery Store   
316             LONDON               4           Supermarket   
318             LONDON               4         Grocery Store   
356             LONDON               4        Discount Store   
359             LONDON               4           Supermarket   
376             LONDON               4         Grocery Store   
441             LONDON               4         Grocery Store   
452             LONDON               4                  Park   
468             LONDON               4         Grocery Store   
517             LONDON               4         Grocery Store   

    2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
1       Indian Restaurant         Train Station                   Park   
9    Fast Food Restaurant                 Hotel          Grocery Store   
29   Fast Food Restaurant                 Hotel          Grocery Store   
61               Pharmacy        Sandwich Place        Warehouse Store   
69               Pharmacy        Sandwich Place        Warehouse Store   
100              Pharmacy        Sandwich Place        Warehouse Store   
127  Fast Food Restaurant                 Hotel          Grocery Store   
137              Pharmacy        Sandwich Place        Warehouse Store   
217              Pharmacy        Sandwich Place        Warehouse Store   
260              Pharmacy        Sandwich Place        Warehouse Store   
267                   Pub                  Café   Brazilian Restaurant   
269              Pharmacy        Sandwich Place        Warehouse Store   
316  Fast Food Restaurant                 Hotel          Grocery Store   
318              Pharmacy        Sandwich Place        Warehouse Store   
356       Warehouse Store         Grocery Store         Sandwich Place   
359  Fast Food Restaurant                 Hotel          Grocery Store   
376              Pharmacy        Sandwich Place        Warehouse Store   
441              Pharmacy        Sandwich Place        Warehouse Store   
452           Supermarket     Indian Restaurant  Portuguese Restaurant   
468           Pizza Place           Coffee Shop     Italian Restaurant   
517              Pharmacy        Sandwich Place        Warehouse Store   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
1          Breakfast Spot            Food Court     Food & Drink Shop   
9       Convenience Store                  Park           Golf Course   
29      Convenience Store                  Park           Golf Course   
61         Discount Store     Convenience Store     Indian Restaurant   
69         Discount Store     Convenience Store     Indian Restaurant   
100        Discount Store     Convenience Store     Indian Restaurant   
127     Convenience Store                  Park           Golf Course   
137        Discount Store     Convenience Store     Indian Restaurant   
217        Discount Store     Convenience Store     Indian Restaurant   
260        Discount Store     Convenience Store     Indian Restaurant   
267           Supermarket           Pizza Place         Train Station   
269        Discount Store     Convenience Store     Indian Restaurant   
316     Convenience Store      

Cluster 5

In [49]:
london_data_nonan.loc[london_data_nonan['Cluster Labels'] == 5, london_data_nonan.columns[[1] + list(range(5, london_data_nonan.shape[1]))]]

town  Cluster Labels 1st Most Common Venue  \
167  LONDON, WELLING               5              Bus Stop   
457    LONDON, ERITH               5              Bus Stop   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
167           Sports Club     Convenience Store           Golf Course   
457           Sports Club     Convenience Store           Golf Course   

    5th Most Common Venue       6th Most Common Venue 7th Most Common Venue  \
167         Historic Site  Construction & Landscaping                  Park   
457         Historic Site  Construction & Landscaping                  Park   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
167               Daycare     Food & Drink Shop            Flower Shop  
457               Daycare     Food & Drink Shop            Flower Shop

## Results and Discussion¶
The neighbourhoods of London are very mulitcultural. There are a lot of different cusines including Indian, Italian, Turkish and Chinese. London seems to take a step further in this direction by having a lot of Restaurants, bars, juice bars, coffee shops, Fish and Chips shop and Breakfast spots. It has a lot of shopping options too with that of the Flea markets, flower shops, fish markets, Fishing stores, clothing stores. The main modes of transport seem to be Buses and trains. For leisure, the neighbourhoods are set up to have lots of parks, golf courses, zoo, gyms and Historic sites.

Overall, the city of London offers a multicultural, diverse and certainly an entertaining experience.

Paris is relatively small in size geographically. It has a wide variety of cusines and eateries including French, Thai, Cambodian, Asian, Chinese etc. There are a lot of hangout spots including many Restaurants and Bars. Paris has a lot of Bistro's. Different means of public transport in Paris which includes buses, bikes, boats or ferries. For leisure and sight seeing, there are a lot of Plazas, Trails, Parks, Historic sites, clothing shops, Art galleries and Museums. Overall, Paris seems like the relaxing vacation spot with a mix of lakes, historic spots and a wide variety of cusines to try out.

## Conclusion
The purpose of this project was to explore the cities of London and Paris and see how attractive it is to potential tourists and migrants. We explored both the cities based on their postal codes and then extrapolated the common venues present in each of the neighbourhoods finally concluding with clustering similar neighbourhoods together.

We could see that each of the neighbourhoods in both the cities have a wide variety of experiences to offer which is unique in it's own way. The cultural diversity is quite evident which also gives the feeling of a sense of inclusion.

Both Paris and London seem to offer a vacation stay or a romantic gateaway with a lot of places to explore, beautiful landscapes and a wide variety of culture.Overall, it's upto the stakeholders to decide which experience they would prefer more and which would more to their liking.